### ISSUE TITLE
Have a License

### ISSUE THREAD

**Issue by harmathy** [crazii/SBEMU#37]
You should choose a license, add a licence file (`COPYING` or `LICENSE` or similar) and license headers to code files. Since you incorporated GPL code, that license would be a natural choice (IANAL, but I guess, you would be obligated to have that license on your code)

**Comment by volkertb** [comment]
The project should indeed have a proper open source license. However, there is currently an [issue](https://github.com/sbemu-x/sbemu-x/issues/4) with the sound driver code that SBEMU uses, which is copied from the [Mpxplay](https://mpxplay.sourceforge.net/) project, which has some kind of restrictive clause that states that the code cannot be used for making alternative versions or forks of Mpxplay or something, which obviously makes that code non-free. (Some of the sound sources appear to have been copied from the Linux ALSA driver sources, some files still having an GPL2 license statement, so having this clause in the source code of a project that apparently partially uses GPL2 sources is problematic in and of itself, but that's another matter.)

Long story short: as long as there isn't any clarity about the sound driver sources in the SBEMU project, I don't think a license can be chosen yet. On the other hand, SBEMU also uses code from the DOSBox project, which is is GPL2-licensed, so that that brings obligations with it as well.

Either the sound sources in this project should be replaced, or verification needs to be done that the sound sources that were copied from the Mpxplay project were in fact GPL2 (or possibly otherwise OSI-licensed) sourced copied by the Mpxplay developer from elsewhere.

**Comment by thp** [comment]
See also:

- https://github.com/sbemu-x/sbemu-x/issues/4#issuecomment-1861822484
- https://sourceforge.net/p/mpxplay/discussion/219198/thread/ac6560662b/

**Comment by crazii** [comment]
There's LGPL 2.1 code used (gormcb.c from DJGPP source code, modified), so which is the compatible license to use? I guess it should be GPLv2?

**Comment by volkertb** [comment]
LGPL 2.1 source code may be used with GPLv2 code. However, unless the LGPL 2.1 code is in a separate dynamically linked library, it will effectively become GPLv2 code:

> 3. You may opt to apply the terms of the ordinary GNU General Public License instead of this License to a given copy of the Library. To do this, you must alter all the notices that refer to this License, so that they refer to the ordinary GNU General Public License, version 2, instead of to this License. (If a newer version than version 2 of the ordinary GNU General Public License has appeared, then you can specify that version instead if you wish.) Do not make any other change in these notices.

Source: https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html#SEC1

**Comment by crazii** [comment]
If this is true, why would DJ states in the `copying.dj` that the source to build libc.a is LGPL? I'm confused. The djgpp libc.a is not a dynamic lib itself. Does that matter?

**Comment by volkertb** [comment]
The most popular C library in the Linux world, glibc, is licensed under the LPGL, and it can be either statically or dynamically linked. The `.a` extension would indeed make it seem that the library was linked statically.

As I understand it, when the code is linked dynamically, it can be used with non-GPL software.

If the code is linked statically, the code that is is linked with must be licensed under the GPL, or a license that is compatible with the GPL.

So yes, the source that is used to build `libc.a` is licensed under the LGPL, since it can also be used to build a DLL.

**Comment by volkertb** [comment]
Let me put it differently: even if we are going to link SBEMU statically with `libc.a`, it will be fine, since SBEMU is licensed under the GPLv2 license.

**Comment by volkertb** [comment]
Oh, wait: apparently the C Library that comes with DJGPP (`libc.a`) has an even less restricted license than GPL:

> Using the GNU C/C++ compiler doesn't make your programs subject to any restrictions. The C library which comes with DJGPP is free (unless you change the library sources, see below), which means you are free to use the stock libc.a in any way you like (but please try to comply with [basic rules of courtesy](https://www.delorie.com/djgpp/v2faq/faq19_2.html).)

Source (including more info on this): https://www.delorie.com/djgpp/v2faq/faq19_1.html

**Comment by volkertb** [comment]
But even if the source code of a library is LGPL, and you link it statically, you may still do so as long as you are only linking it statically with source code that is either licensed under the GPL or has a license that is compatible with the GPL.

LGPL only affords additional flexibility (the right to link it with non-GPL or even closed-source software) when you link the LGPL code as a dynamic (.DLL) or shared (.so) library.

**Comment by 4nd3r50ncr** [comment]
Now we have a license. GNU Version 2. Follow the [SBEMU](https://github.com/crazii/SBEMU/) main page.

**Comment by volkertb** [comment]
@4nd3r50ncr @crazii Thanks! I did however notice that the `COPYING` file in this project has the `or later` clause. I'm not sure if that's possible, since I believe that some of the sources from other open source projects that we rely on is licensed under GPLv2 without the optional "or later" clause.

Software that is licensed as `GPLv2 or any later version` can be combined with GPLv2 or GPLv3 project, even if the existing code in that project doesn't have the "or later" clause, in which case the code base as a whole will effectively not have the "or later" clause.

The other way around, however, including GPL code without "or later" in a project that is licensed with the "or later" clause is not permitted, unless the "or later" clause is removed for that entire project.

I know, the nitty-gritty of open source licenses is really a frustrating topic sometimes, but please don't shoot the messenger. :flushed:

Shall I open a separate issue for this? Or shall I reopen this issue?

**Comment by 4nd3r50ncr** [comment]
> @4nd3r50ncr @crazii Thanks! I did however notice that the `COPYING` file in this project has the `or later` clause. I'm not sure if that's possible, since I believe that some of the sources from other open source projects that we rely on is licensed under GPLv2 without the optional "or later" clause.
> 
> Software that is licensed as `GPLv2 or any later version` can be combined with GPLv2 or GPLv3 project, even if the existing code in that project doesn't have the "or later" clause, in which case the code base as a whole will effectively not have the "or later" clause.
> 
> The other way around, however, including GPL code without "or later" in a project that is licensed with the "or later" clause is not permitted, unless the "or later" clause is removed for that entire project.
> 
> I know, the nitty-gritty of open source licenses is really a frustrating topic sometimes, but please don't shoot the messenger. 😳
> 
> Shall I open a separate issue for this? Or shall I reopen this issue?

I believe in your experience, @volkertb feel free.

**Comment by crazii** [comment]
Cool, you can open an issue or, just remove the 'or later' phrase in COPYING, does that work?

**Comment by volkertb** [comment]
@crazii Will do, once the release pipeline issues have been resolved, and it's safe to merge stuff again. :slightly_smiling_face:
